libs/wampcc/wamp_router.cc

```diff
@@ -264,15 +289,38 @@ std::future<uverr> wamp_router::listen(auth_provider auth,
                               json_object& details,
                               wamp_args& args)
     {
-      ws.authorize(uri, auth_provider::action::call);
       this->handle_inbound_call(&ws, reqid, uri, details, args);
     };
 
     handlers.on_publish =
     [this](wamp_session& ws, t_request_id request_id, std::string uri,
            json_object options, wamp_args args) {
-      ws.authorize(uri, auth_provider::action::publish);
       try {
+        /* Check if this publish is authorized */
+        auto authorization = ws.authorize(uri, auth_provider::action::publish);
+
+        if( !authorization.allow )
+          throw wamp_error(WAMP_ERROR_NOT_AUTHORIZED, "publish is not authorized");
+
+        json_object details;
+
+        if( authorization.disclose ) {
+          /* Populate publisher session details */
+          details["caller"] = ws.unique_id();
+          if(ws.has_authid())
+            details["caller_authid"] = ws.authid();
+          details["caller_authrole"] = ws.authrole();
+
```

Comment:
[14.3.5.1](https://wamp-proto.org/static/rfc/draft-oberstet-hybi-crossbar-wamp.html#rfc.section.14.3.5.1) Feature Definition
> Note that a Dealer MAY disclose the identity of a Caller even without the Caller having explicitly requested to do so when the Dealer configuration (for the called procedure) is setup to do so.

This would mean that the dealer has the final work on whether to disclose the identity or not.

But maybe it would make sense to allow disclosure configuration to be optional instead of `true / false`
```
enum disclosure {
    optional,       // sends caller-id only if requested by caller
    always,         // always disclose caller-id
    never           // never disclose caller-id
};
```
There is also an option for the callee to request `{"disclose_caller":true}`. But there is no description on how to resolve dealer+caller disclosure settings and what error to return on failure.
There is no `wamp.error.disclose_caller.not_allowed` error specified yet.

Reply:
yeah, I think having that disclosure enum is sensible